File: android-compose/app/src/main/java/com/escom/tarea2/compose/ui/screens/ContainersScreen.kt
package com.escom.tarea2.compose.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.GridView
import androidx.compose.material.icons.filled.Home
import androidx.compose.material.icons.filled.List
import androidx.compose.material.icons.filled.Settings
import androidx.compose.material3.Button
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.NavigationBar
import androidx.compose.material3.NavigationBarItem
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.escom.tarea2.compose.ui.components.CatalogItemCard
import com.escom.tarea2.compose.viewmodel.SharedCatalogViewModel

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ContainersScreen(viewModel: SharedCatalogViewModel) {
    var selectedNavIndex by remember { mutableIntStateOf(0) }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(16.dp)
    ) {
        Text(
            text = "Sección 6: Contenedores y Estructura",
            style = MaterialTheme.typography.headlineMedium,
            color = MaterialTheme.colorScheme.primary
        )
        Text(
            text = "Disposición espacial bidimensional en Compose mediante Row, Column, Box, ScrollView, TopAppBar, NavigationBar y pesos Modifier.weight.",
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )

        Spacer(modifier = Modifier.height(12.dp))

        // 1. Distribución en Fila, Columna y Superposición (Box)
        CatalogItemCard(
            title = "1. Distribución en Fila (Row), Columna (Column) y Superpuesta (Box)",
            description = "Row alinea elementos horizontalmente, Column verticalmente y Box apila elementos en el eje Z con modificadores de alineación."
        ) {
            Column {
                Text("Row (Horizontal):", style = MaterialTheme.typography.labelMedium)
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(vertical = 6.dp)
                ) {
                    Box(
                        modifier = Modifier
                            .weight(1f)
                            .height(44.dp)
                            .background(MaterialTheme.colorScheme.primaryContainer, RoundedCornerShape(8.dp)),
                        contentAlignment = Alignment.Center
                    ) {
                        Text("Fila 1 (50%)", color = MaterialTheme.colorScheme.onPrimaryContainer)
                    }
                    Spacer(modifier = Modifier.padding(4.dp))
                    Box(
                        modifier = Modifier
                            .weight(1f)
                            .height(44.dp)
                            .background(MaterialTheme.colorScheme.secondaryContainer, RoundedCornerShape(8.dp)),
                        contentAlignment = Alignment.Center
                    ) {
                        Text("Fila 2 (50%)", color = MaterialTheme.colorScheme.onSecondaryContainer)
                    }
                }

                Spacer(modifier = Modifier.height(8.dp))

                Text("Box Superpuesto (Stack Z-Order):", style = MaterialTheme.typography.labelMedium)
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(90.dp)
                        .background(MaterialTheme.colorScheme.surfaceVariant, RoundedCornerShape(10.dp))
                        .padding(8.dp)
                ) {
                    Text(
                        text = "Capa Inferior de Fondo",
                        modifier = Modifier.align(Alignment.Center)
                    )
                    Button(
                        onClick = { viewModel.setLastAction("Pulsado botón flotante en Box") },
                        modifier = Modifier.align(Alignment.BottomEnd)
                    ) {
                        Text("Capa Superior")
                    }
                }
            }
        }

        // 2. Pesos proporcionales
        CatalogItemCard(
            title = "2. Pesos Proporcionales (Modifier.weight)",
            description = "Reparte el ancho disponible de manera proporcional (1:2:1) calculando automáticamente las dimensiones de cada nodo hijo."
        ) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(48.dp)
            ) {
                Box(
                    modifier = Modifier
                        .weight(1f)
                        .fillMaxSize()
                        .background(MaterialTheme.colorScheme.primary, RoundedCornerShape(topStart = 8.dp, bottomStart = 8.dp)),
                    contentAlignment = Alignment.Center
                ) {
                    Text("Peso 1 (25%)", color = Color.White, style = MaterialTheme.typography.labelSmall)
                }
                Box(
                    modifier = Modifier
                        .weight(2f)
                        .fillMaxSize()
                        .background(MaterialTheme.colorScheme.secondary),
                    contentAlignment = Alignment.Center
                ) {
                    Text("Peso 2 (50%)", color = Color.White, style = MaterialTheme.typography.labelSmall)
                }
                Box(
                    modifier = Modifier
                        .weight(1f)
                        .fillMaxSize()
                        .background(MaterialTheme.colorScheme.tertiary, RoundedCornerShape(topEnd = 8.dp, bottomEnd = 8.dp)),
                    contentAlignment = Alignment.Center
                ) {
                    Text("Peso 1 (25%)", color = Color.White, style = MaterialTheme.typography.labelSmall)
                }
            }
        }

        // 3. Barra Superior y Barra Inferior
        CatalogItemCard(
            title = "3. Estructuras Globales: TopAppBar y NavigationBar",
            description = "Scaffolding de Material 3 para barras superiores de aplicación y navegación inferior interactiva entre destinos clave."
        ) {
            Column {
                TopAppBar(
                    title = { Text("TopAppBar de Muestra", style = MaterialTheme.typography.titleSmall) },
                    navigationIcon = {
                        Icon(Icons.Default.GridView, contentDescription = null, modifier = Modifier.padding(start = 8.dp))
                    },
                    colors = TopAppBarDefaults.topAppBarColors(
                        containerColor = MaterialTheme.colorScheme.primaryContainer,
                        titleContentColor = MaterialTheme.colorScheme.onPrimaryContainer
                    )
                )

                Spacer(modifier = Modifier.height(10.dp))

                NavigationBar {
                    val navItems = listOf(
                        Triple("Inicio", Icons.Default.Home, 0),
                        Triple("Colección", Icons.Default.List, 1),
                        Triple("Ajustes", Icons.Default.Settings, 2)
                    )

                    navItems.forEach { (label, icon, index) ->
                        NavigationBarItem(
                            selected = selectedNavIndex == index,
                            onClick = {
                                selectedNavIndex = index
                                viewModel.setLastAction("NavigationBar pulsada: $label")
                            },
                            icon = { Icon(icon, contentDescription = null) },
                            label = { Text(label) }
                        )
                    }
                }
            }
        }
    }
}

File: flutter/lib/screens/containers_screen.dart
import 'package:flutter/material.dart';
import '../state/catalog_state.dart';
import '../widgets/catalog_card.dart';

class ContainersScreen extends StatefulWidget {
  final CatalogState state;

  const ContainersScreen({super.key, required this.state});

  @override
  State<ContainersScreen> createState() => _ContainersScreenState();
}

class _ContainersScreenState extends State<ContainersScreen> {
  int _miniNavIndex = 0;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);

    return SingleChildScrollView(
      padding: const EdgeInsets.all(16.0),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(
            'Sección 6: Contenedores y Estructura',
            style: theme.textTheme.headlineMedium?.copyWith(
              color: theme.colorScheme.primary,
              fontWeight: FontWeight.bold,
            ),
          ),
          Text(
            'Esquemas de composición bidimensional mediante Row, Column, Stack, SingleChildScrollView, AppBar y flex proporcionales.',
            style: theme.textTheme.bodyMedium?.copyWith(
              color: theme.colorScheme.onSurfaceVariant,
            ),
          ),
          const SizedBox(height: 12),

          // 1. Fila, Columna y Superposición (Stack)
          CatalogCard(
            title: '1. Distribución en Fila (Row), Columna (Column) y Superpuesta (Stack)',
            description: 'Row dispone hijos sobre el eje horizontal, Column sobre el vertical y Stack los apila en el eje Z permitiendo posiciones absolutas (Positioned).',
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                const Text('Row (Horizontal):', style: TextStyle(fontWeight: FontWeight.w600)),
                const SizedBox(height: 6),
                Row(
                  children: [
                    Expanded(
                      child: Container(
                        height: 40,
                        decoration: BoxDecoration(
                          color: theme.colorScheme.primaryContainer,
                          borderRadius: BorderRadius.circular(8),
                        ),
                        alignment: Alignment.center,
                        child: Text(
                          'Fila 1 (50%)',
                          style: TextStyle(color: theme.colorScheme.onPrimaryContainer),
                        ),
                      ),
                    ),
                    const SizedBox(width: 8),
                    Expanded(
                      child: Container(
                        height: 40,
                        decoration: BoxDecoration(
                          color: theme.colorScheme.secondaryContainer,
                          borderRadius: BorderRadius.circular(8),
                        ),
                        alignment: Alignment.center,
                        child: Text(
                          'Fila 2 (50%)',
                          style: TextStyle(color: theme.colorScheme.onSecondaryContainer),
                        ),
                      ),
                    ),
                  ],
                ),
                const SizedBox(height: 12),
                const Text('Stack (Superposición):', style: TextStyle(fontWeight: FontWeight.w600)),
                const SizedBox(height: 6),
                SizedBox(
                  height: 90,
                  width: double.infinity,
                  child: Stack(
                    children: [
                      Container(
                        width: double.infinity,
                        height: double.infinity,
                        decoration: BoxDecoration(
                          color: theme.colorScheme.surfaceContainerHighest,
                          borderRadius: BorderRadius.circular(8),
                        ),
                        alignment: Alignment.center,
                        child: const Text('Capa Inferior de Fondo'),
                      ),
                      Positioned(
                        right: 8,
                        bottom: 8,
                        child: FilledButton.tonal(
                          onPressed: () {
                            widget.state.setLastAction('Pulsado botón flotante en Stack');
                          },
                          child: const Text('Capa Flotante'),
                        ),
                      ),
                    ],
                  ),
                ),
              ],
            ),
          ),

          // 2. Pesos proporcionales (Expanded con flex)
          CatalogCard(
            title: '2. Pesos Proporcionales (Expanded con flex)',
            description: 'Divide el espacio disponible en proporciones exactas mediante la propiedad flex (1:2:1 = 25% : 50% : 25%).',
            child: Row(
              children: [
                Expanded(
                  flex: 1,
                  child: Container(
                    height: 48,
                    decoration: BoxDecoration(
                      color: theme.colorScheme.primary,
                      borderRadius: const BorderRadius.horizontal(left: Radius.circular(8)),
                    ),
                    alignment: Alignment.center,
                    child: const Text('Flex 1 (25%)', style: TextStyle(color: Colors.white, fontSize: 11)),
                  ),
                ),
                Expanded(
                  flex: 2,
                  child: Container(
                    height: 48,
                    color: theme.colorScheme.secondary,
                    alignment: Alignment.center,
                    child: const Text('Flex 2 (50%)', style: TextStyle(color: Colors.white, fontSize: 11)),
                  ),
                ),
                Expanded(
                  flex: 1,
                  child: Container(
                    height: 48,
                    decoration: BoxDecoration(
                      color: theme.colorScheme.tertiary,
                      borderRadius: const BorderRadius.horizontal(right: Radius.circular(8)),
                    ),
                    alignment: Alignment.center,
                    child: const Text('Flex 1 (25%)', style: TextStyle(color: Colors.white, fontSize: 11)),
                  ),
                ),
              ],
            ),
          ),

          // 3. Barra Superior e Inferior Demostrativas
          CatalogCard(
            title: '3. Barras de Estructura: Superior (AppBar) e Inferior (NavigationBar)',
            description: 'Widgets estructurales que delimitan las áreas de título, acciones de cabecera y selección de rutas principales.',
            child: Column(
              children: [
                Container(
                  padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 10),
                  decoration: BoxDecoration(
                    color: theme.colorScheme.primaryContainer,
                    borderRadius: BorderRadius.circular(8),
                  ),
                  child: Row(
                    mainAxisAlignment: MainAxisAlignment.spaceBetween,
                    children: [
                      Row(
                        children: [
                          Icon(Icons.grid_view, color: theme.colorScheme.onPrimaryContainer),
                          const SizedBox(width: 12),
                          Text(
                            'AppBar de Muestra',
                            style: TextStyle(
                              color: theme.colorScheme.onPrimaryContainer,
                              fontWeight: FontWeight.bold,
                            ),
                          ),
                        ],
                      ),
                      Icon(Icons.more_vert, color: theme.colorScheme.onPrimaryContainer),
                    ],
                  ),
                ),
                const SizedBox(height: 10),
                NavigationBar(
                  selectedIndex: _miniNavIndex,
                  onDestinationSelected: (idx) {
                    setState(() => _miniNavIndex = idx);
                    final titles = ['Inicio', 'Colección', 'Ajustes'];
                    widget.state.setLastAction('NavigationBar seleccionada: ${titles[idx]}');
                  },
                  destinations: const [
                    NavigationDestination(icon: Icon(Icons.home), label: 'Inicio'),
                    NavigationDestination(icon: Icon(Icons.list), label: 'Colección'),
                    NavigationDestination(icon: Icon(Icons.settings), label: 'Ajustes'),
                  ],
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}
